docs/csharp/language-reference/keywords/lock-statement.md

```diff
@@ -11,76 +11,69 @@ ms.assetid: 656da1a4-707e-4ef6-9c6e-6d13b646af42
-            {
-                throw new Exception("Insufficient funds");
-            }
-            balance -= amount;
-        }
-    }
+    // Your code...
 }
 ```
 
-For more information, see [Thread Synchronization](../../programming-guide/concepts/threading/thread-synchronization.md).
-
-## Remarks
-
-The `lock` keyword ensures that one thread does not enter a critical section of code while another thread is in the critical section. If another thread tries to enter a locked code, it will wait, block, until the object is released.
+where `x` is an expression of a [reference type](reference-types.md) is precisely equivalent to
 
-The section [Threading](../../programming-guide/concepts/threading/index.md) discusses threading.
+```csharp
+bool __lockWasTaken = false;
+try
+{
+    System.Threading.Monitor.Enter(x, ref __lockWasTaken);
+    // Your code...
+}
+finally
+{
+    if (__lockWasTaken) System.Threading.Monitor.Exit(x);
+}
+```
 
-The `lock` keyword calls <xref:System.Threading.Monitor.Enter%2A> at the start of the block and <xref:System.Threading.Monitor.Exit%2A> at the end of the block. A <xref:System.Threading.ThreadInterruptedException> is thrown if <xref:System.Threading.Thread.Interrupt%2A> interrupts a thread that is waiting to enter a `lock` statement.
+except that `x` is only evaluated once.
 
-In general, avoid locking on a `public` type, or instances beyond your code's control. The common constructs `lock (this)`, `lock (typeof (MyType))`, and `lock ("myLock")` violate this guideline:
+Since the code uses a [try...finally](try-finally.md) block, the lock is released even if an exception is thrown within the body of a `lock` statement.
 
-- `lock (this)` is a problem if the instance can be accessed publicly.
+While a lock is held, the same thread can again obtain and release the lock. However, any other thread is blocked from obtaining the lock and waits until the lock is released.
 
-- `lock (typeof (MyType))` is a problem if `MyType` is publicly accessible.
+You can't use the [await](await.md) keyword in the body of a `lock` statement.
 
-- `lock("myLock")` is a problem because any other code in the process using the same string, will share the same lock.
+## Remarks
 
-Best practice is to define a `private` object to lock on, or a `private static` object variable to protect data common to all instances.
+Avoid locking on publicly accessible instances, as it might result in lock contention or deadlock. Consider the following constructs:
 
-You can't use the [await](await.md) keyword in the body of a `lock` statement.
+- `lock(this)` is a problem if the instance referenced by `this` can be accessed publicly.
 
-## Example - Threads without locking
+- `lock(typeof(MyType))` is a problem if `MyType` is publicly accessible.
 
-The following sample shows a simple use of threads without locking in C#:
+- `lock("myLock")` is a problem because any other code in the process using the same string, would share the same lock.
```

Comment:
This applies to all string instances to do posibility of string interning. It is also possible go in the other direction and end up not sharing a lock when you want it - AFAIR it's possible that the same string literal to yield different string instances in different assemblies.